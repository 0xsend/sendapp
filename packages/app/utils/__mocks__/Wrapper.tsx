import { QueryClient, QueryClientConfig, QueryClientProvider } from '@tanstack/react-query'
import { useState } from 'react'

export function Wrapper({
  children,
  defaultOptions,
}: { children: React.ReactNode; defaultOptions?: QueryClientConfig['defaultOptions'] }) {
  const [queryClient] = useState(
    () =>
      new QueryClient({
        defaultOptions: {
          ...defaultOptions,
          queries: {
            retry: false,
            gcTime: Infinity,
            ...defaultOptions?.queries,
          },
        },
      })
  )
  return <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
}
