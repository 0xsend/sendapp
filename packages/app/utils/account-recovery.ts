import { createSupabaseAdminClient } from 'app/utils/supabase/admin'
import type { Debugger } from 'debug'

/**
 * Retrieve a chainAddress row from the `chain_addresses` table
 * @param {string} chainAddress - chainAddress to retrieve
 * @returns chainAddress row
 */
export const getChainAddress = async (chainAddress: `0x${string}`) => {
  const supabaseAdmin = createSupabaseAdminClient()
  return await supabaseAdmin
    .from('chain_addresses')
    .select('*')
    .eq('address', chainAddress)
    .single()
}

/**
 * Retrieves a passkey row from the `webauthn_credentials` table by its name `passkeyName` in the format <userId>.<keySlot>
 * @param {string} passkeyName - passkey name, in the format <userId>.<keySlot>
 * @returns passkey row
 */
export const getPasskey = async (passkeyName: string) => {
  const supabaseAdmin = createSupabaseAdminClient()
  return await supabaseAdmin
    .from('webauthn_credentials')
    .select('*')
    .eq('name', passkeyName)
    .single()
}

/**
 * Retrieves a challenge from the `challenges` table, from the challenge id.
 * @param {number} challengeId
 * @returns challenge row
 */
export const getChallengeById = async (challengeId: number) => {
  const supabaseAdmin = createSupabaseAdminClient()
  return await supabaseAdmin.from('challenges').select('*').eq('id', challengeId).single()
}

/**
 * Determines whether a challenge is expired
 * @param {number} challengeId - challenge id in `challenges` table
 * @param {Debugger} logger - logger
 * @returns {boolean} - whether the challenge is expired
 */
export const isChallengeExpired = async (
  challengeId: number,
  logger?: Debugger
): Promise<boolean> => {
  const supabaseAdmin = createSupabaseAdminClient()
  const { data, error } = await supabaseAdmin
    .from('challenges')
    .select('*')
    .eq('id', challengeId)
    .gt('expires_at', 'now()')
    .limit(1)

  if (error) {
    logger?.(`isChallengeExpired:${error?.message}`)
    throw error
  }

  if (data.length === 0) {
    return true
  }
  return false
}
