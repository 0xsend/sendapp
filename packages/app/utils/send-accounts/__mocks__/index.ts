const mockSendAccounts = {
  useSendAccounts: jest.fn().mockReturnValue({
    data: [
      {
        webauthn_credentials: [
          {
            public_key: '\\x1234567890',
          },
        ],
      },
    ],
  }),
}

export default mockSendAccounts
