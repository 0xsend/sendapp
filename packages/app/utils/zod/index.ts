export {
  byteaEthAddress,
  byteaToHexEthAddress,
  byteaTxHash,
  byteaToHexTxHash,
} from './bytea'
export { decimalStrToBigInt } from './bigint'
export { hex, address } from './evm'
export { epochToDate } from './activity/time'
