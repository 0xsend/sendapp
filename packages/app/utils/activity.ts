import { formatUnits } from 'viem'
import formatAmount from './formatAmount'
import { isTagReceiptsEvent, isReferralsEvent, isSendAccountTransfersEvent } from './zod/activity'
import type { Activity } from 'app/utils/zod/activity'
import { isSendAccountReceiveEvent } from './zod/activity/SendAccountReceiveEventSchema'
import { tokenPaymasterAddress } from '@my/wagmi'

const wagmiAddresWithLabel = (addresses: `0x${string}`[], label: string) =>
  Object.values(addresses).map((a) => [a, label])

const AddressLabels = {
  ...Object.fromEntries(wagmiAddresWithLabel(Object.values(tokenPaymasterAddress), 'Paymaster')),
}

const labelAddress = (address: `0x${string}`): string => AddressLabels[address] ?? address

/**
 * Returns the counterpart of the activity which could be the logged in user.
 * If the activity is a send account transfer or receive,
 *   if received, the counterpart is the user who sent the token.
 *   if sent, the counterpart is the user who received the token.
 * If the activity is a tag receipt, the actor is the user who created the tag.
 * If the activity is a referral, the actor is the user who referred the user.
 */
export function counterpart(activity: Activity): Activity['from_user'] | Activity['to_user'] {
  const { from_user, to_user } = activity
  if (isTagReceiptsEvent(activity)) {
    return from_user
  }
  if (isReferralsEvent(activity) && !!from_user?.id) {
    return to_user // show the referred
  }
  if (isReferralsEvent(activity) && !!to_user?.id) {
    return from_user // show the referrer
  }
  if (isSendAccountTransfersEvent(activity) || isSendAccountReceiveEvent(activity)) {
    if (from_user?.id) {
      // if i am the sender, show the receiver
      return to_user
    }
    if (to_user?.id) {
      // if i am the receiver, show the sender
      return from_user
    }
  }
  return null // not a send or receive event
}

/**
 * Returns the amount of the activity if there is one.
 */
export function amountFromActivity(activity: Activity): string {
  switch (true) {
    case isSendAccountTransfersEvent(activity): {
      const { v, coin } = activity.data
      if (coin) {
        const amount = formatUnits(v, coin.decimals)
        return `${amount} ${coin.symbol}`
      }
      return formatAmount(`${v}`, 5, 0)
    }
    case isSendAccountReceiveEvent(activity): {
      const { coin } = activity.data
      if (coin) {
        const amount = formatUnits(activity.data.value, coin.decimals)
        return `${amount} ${coin.symbol}`
      }
      return formatAmount(`${activity.data.value}`, 5, 0)
    }
    case isTagReceiptsEvent(activity): {
      const data = activity.data
      const amount = formatUnits(data.value, data.coin.decimals)
      return `${amount} ${data.coin.symbol}`
    }
    case isReferralsEvent(activity) && !!activity.from_user?.id: {
      // only show if the user is the referrer
      const data = activity.data
      return `${data.tags.length} Referrals`
    }
    case isReferralsEvent(activity) && !!activity.to_user?.id: {
      // only show if the user is the referred
      return '' // no amount
    }
    case activity.event_name === 'send_account_signing_key_added': {
      return ''
    }
    case activity.event_name === 'send_account_signing_key_removed': {
      return ''
    }
    default:
      if (__DEV__) console.log('unknown activity', activity)
      return ''
  }
}

/**
 * Returns the human readable event name of the activity.
 * @param activity
 * @returns
 */
export function eventNameFromActivity(activity: Activity) {
  const { event_name, from_user, to_user } = activity
  const isTransferOrReceive =
    isSendAccountTransfersEvent(activity) || isSendAccountReceiveEvent(activity)
  switch (true) {
    case isTransferOrReceive && !!to_user?.id:
      return 'Received'
    case isTransferOrReceive && !!from_user?.id:
      return 'Sent'
    case isTagReceiptsEvent(activity):
      return 'Sendtag Registered'
    case isReferralsEvent(activity) && !!from_user?.id:
      return 'Referral'
    case isReferralsEvent(activity) && !!to_user?.id:
      return 'Referred By'
    default:
      return event_name // catch-all i_am_rick_james -> I Am Rick James
        .split('_')
        .map((s) => s.charAt(0).toUpperCase() + s.slice(1))
        .join(' ')
  }
}

/**
 * Returns the subtext of the activity if there is one.
 */
export function subtextFromActivity(activity: Activity): string | null {
  const _user = counterpart(activity)
  const { from_user, to_user } = activity
  if (isTagReceiptsEvent(activity)) {
    return activity.data.tags.map((t) => `/${t}`).join(', ')
  }
  if (isReferralsEvent(activity)) {
    if (from_user?.id) {
      // show the referred
      return userNameFromActivityUser(to_user)
    }
    // show the referrer
    return userNameFromActivityUser(from_user)
  }
  if (_user) {
    return userNameFromActivityUser(_user)
  }
  if (isSendAccountTransfersEvent(activity) && from_user?.id) {
    return labelAddress(activity.data.t)
  }
  if (isSendAccountReceiveEvent(activity) && from_user?.id) {
    return labelAddress(activity.data.sender)
  }
  if (isSendAccountTransfersEvent(activity) && to_user?.id) {
    return labelAddress(activity.data.f)
  }
  if (isSendAccountReceiveEvent(activity) && to_user?.id) {
    return labelAddress(activity.data.log_addr)
  }
  return null
}

/**
 * Returns the name of the user from the activity user.
 * The cascading fallback is to:
 * 1. First, sendtag
 * 2. 2nd, profile name
 * 3. 3rd, Send ID
 */
export function userNameFromActivityUser(
  user: Activity['from_user'] | Activity['to_user']
): string {
  switch (true) {
    case !!user?.tags?.[0]:
      return `/${user.tags[0]}`
    case !!user?.name:
      return user.name
    case !!user?.send_id:
      return `#${user.send_id}`
    case user === null:
      return ''
    default:
      console.error('no user name found', user)
      if (__DEV__) {
        throw new Error('no user name found')
      }
      return ''
  }
}
