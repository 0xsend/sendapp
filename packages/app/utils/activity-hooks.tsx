/**
 * Activity hooks is a new version of the activity utils. This file was created to
 * avoid large conflicts due to the introduction of the activity hooks and
 * the address book.
 *
 * The activity hooks are capapble of returning React components handling more
 * complex cases when rendering an activity. This is useful for cases like
 * Send Earn deposits and withdrawals, where the amount is calculated based on
 * the underlying vault or the addresses are only known in the address book.
 *
 * Once `dev` and the activity feed is stable, this file can be consolidated
 * into the `activity` file again.
 */

import { SendEarnAmount } from 'app/features/earn/components/SendEarnAmount'
import type { ReactNode } from 'react'
import { amountFromActivity } from './activity'
import { type Activity, isSendEarnEvent } from './zod/activity'
import type { LiquidityPool } from './zod/LiquidityPoolSchema'
import type { SwapRouter } from './zod/SwapRouterSchema'

/**
 * Returns the amount of the activity if there is one.
 */
export function useAmountFromActivity(
  activity: Activity,
  swapRouters?: SwapRouter[],
  liquidityPools?: LiquidityPool[]
): ReactNode {
  if (isSendEarnEvent(activity)) {
    return <SendEarnAmount activity={activity} />
  }
  return amountFromActivity(activity, swapRouters, liquidityPools)
}
