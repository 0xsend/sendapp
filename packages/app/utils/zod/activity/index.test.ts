import { describe, expect, it } from '@jest/globals'
import { assert } from 'app/utils/assert'
import { EventArraySchema, type BaseEvent } from '.'
import type { SendAccountTransfersEvent } from './SendAccountTransfersEventSchema'
import type { TagReceiptsEvent } from './TagReceiptsEventSchema'
import { MockActivityFeed } from 'app/features/activity/utils/__mocks__/mock-activity-feed'
import type { ReferralsEvent } from './ReferralsEventSchema'
import type { SendAccountReceiveEvent } from './SendAccountReceiveEventSchema'
import type { TagReceiptUSDCEvent } from './TagReceiptUSDCEventSchema'
import { ethCoin, sendCoin, usdcCoin } from 'app/data/coins'

describe('EventArraySchema', () => {
  let parsedData: BaseEvent[]

  beforeAll(() => {
    const result = EventArraySchema.safeParse(MockActivityFeed)
    if (!result.success) {
      throw new Error(`Failed to parse MockActivityFeed: ${result.error}`)
    }
    parsedData = result.data
  })

  it('should parse a valid event array', () => {
    expect(parsedData).toMatchSnapshot()
    expect(parsedData).toHaveLength(10)
  })

  describe('SendAccountTransfersEvent', () => {
    let transfer: SendAccountTransfersEvent

    beforeAll(() => {
      transfer = parsedData[0] as SendAccountTransfersEvent
      assert(!!transfer)
    })

    it('should have correct event name and timestamps', () => {
      expect(transfer.event_name).toBe('send_account_transfers')
      expect(transfer.created_at).toBeInstanceOf(Date)
    })

    it('should have correct user data', () => {
      expect(transfer.from_user).toBeNull()
      expect(transfer.to_user).not.toBeNull()
    })

    it('should have correct transfer data', () => {
      expect(transfer.data.f).toBe('0x760E2928C3aa3aF87897bE52eb4833d42bbB27cf')
      expect(transfer.data.t).toBe('0xbf65EE06b43B9cA718216241f0b9F81b5ff30CC1')
      expect(transfer.data.v).toBe(19032n)
      expect(transfer.data.block_num).toBe(15164540n)
      expect(transfer.data.tx_idx).toBe(0n)
      expect(transfer.data.log_idx).toBe(2n)
      expect(transfer.data.tx_hash).toBe(
        '0xf1443b5abd14e6212dda2d9f5ff1d1d691599de3e8fa019ccc19b909d9bb46a4'
      )
      expect(transfer.data.coin?.label).toBe('USDC')
    })
  })

  describe('TagReceiptsEvent', () => {
    let receipt: TagReceiptsEvent

    beforeAll(() => {
      receipt = parsedData[1] as TagReceiptsEvent
      assert(!!receipt)
    })

    it('should have correct event name and timestamps', () => {
      expect(receipt.event_name).toBe('tag_receipts')
      expect(receipt.created_at).toBeInstanceOf(Date)
    })

    it('should have correct user data', () => {
      expect(receipt.from_user).not.toBeNull()
      expect(receipt.to_user).toBeNull()
    })

    it('should have correct receipt data', () => {
      expect(receipt.data.tags).toEqual(['yuw'])
      expect(receipt.data.value).toBe(20000000000000000n)
      expect(receipt.data.tx_hash).toBe(
        '0x37c4281422413a3a78e765452c47abb5c3a95c102282bdd3632ced0b640d861c'
      )
      expect(receipt.data.coin).toEqual(ethCoin)
    })
  })

  describe('TagReceiptUSDCEvent', () => {
    let receipt: TagReceiptUSDCEvent

    beforeAll(() => {
      receipt = parsedData[7] as TagReceiptUSDCEvent
      assert(!!receipt)
    })

    it('should have correct event name and timestamps', () => {
      expect(receipt.event_name).toBe('tag_receipt_usdc')
      expect(receipt.created_at).toBeInstanceOf(Date)
    })

    it('should have correct user data', () => {
      expect(receipt.from_user).not.toBeNull()
      expect(receipt.to_user).toBeNull()
    })

    it('should have correct receipt data', () => {
      expect(receipt.data.tags).toEqual(['tag_receipt_usdc'])
      expect(receipt.data.value).toBe(2000000n)
      expect(receipt.data.tx_hash).toBe(
        '0x37c4281422413a3a78e765452c47abb5c3a95c102282bdd3632ced0b640d861c'
      )
      expect(receipt.data.coin).toEqual(usdcCoin)
    })
  })

  describe('ReferralsEvent', () => {
    let referral: ReferralsEvent

    beforeAll(() => {
      referral = parsedData[2] as ReferralsEvent
      assert(!!referral)
    })

    it('should have correct event name and timestamps', () => {
      expect(referral.event_name).toBe('referrals')
      expect(referral.created_at).toBeInstanceOf(Date)
    })

    it('should have correct user data', () => {
      expect(referral.from_user).not.toBeNull()
      expect(referral.to_user).not.toBeNull()
    })

    it('should have correct referral data', () => {
      expect(referral.data.tags).toEqual(['disconnect_whorl7351'])
    })
  })

  describe('SendAccountSigningKeyAddedEvent', () => {
    let keyAdded: BaseEvent

    beforeAll(() => {
      keyAdded = parsedData[3] as BaseEvent
      assert(!!keyAdded)
    })

    it('should have correct event name and timestamps', () => {
      expect(keyAdded.event_name).toBe('send_account_signing_key_added')
      expect(keyAdded.created_at).toBeInstanceOf(Date)
    })

    it('should have correct user data', () => {
      expect(keyAdded.from_user).not.toBeNull()
      expect(keyAdded.to_user).toBeNull()
    })

    it('should have correct key added data', () => {
      expect(keyAdded.data.account).toBe('\\xa7ded3f6316c7d3b5ae2ed711cf535395db921b1')
      expect(keyAdded.data.key_slot).toBe(0)
      expect(keyAdded.data.key).toEqual([
        '\\x351631d94d8cfc12f6adfc2586111990681f216c7d6d8531e669471293f32f07',
        '\\x83577aa62079c3bb5b813017df43832562d133feb3a7447d28849dac74c8aa43',
      ])
    })
  })

  describe('SendAccountSigningKeyRemovedEvent', () => {
    let keyRemoved: BaseEvent

    beforeAll(() => {
      keyRemoved = parsedData[4] as BaseEvent
      assert(!!keyRemoved)
    })

    it('should have correct event name and timestamps', () => {
      expect(keyRemoved.event_name).toBe('send_account_signing_key_removed')
      expect(keyRemoved.created_at).toBeInstanceOf(Date)
    })

    it('should have correct user data', () => {
      expect(keyRemoved.from_user).not.toBeNull()
      expect(keyRemoved.to_user).toBeNull()
    })

    it('should have correct key removed data', () => {
      expect(keyRemoved.data.account).toBe('\\xa7ded3f6316c7d3b5ae2ed711cf535395db921b1')
      expect(keyRemoved.data.key_slot).toBe(0)
      expect(keyRemoved.data.key).toEqual([
        '\\x351631d94d8cfc12f6adfc2586111990681f216c7d6d8531e669471293f32f07',
        '\\x83577aa62079c3bb5b813017df43832562d133feb3a7447d28849dac74c8aa43',
      ])
    })
  })

  describe('SendAccountReceiveEvent', () => {
    let sendAccountReceive: SendAccountReceiveEvent

    beforeAll(() => {
      sendAccountReceive = parsedData[6] as SendAccountReceiveEvent
      assert(!!sendAccountReceive)
    })

    it('should have correct event name and timestamps', () => {
      expect(sendAccountReceive.event_name).toBe('send_account_receives')
      expect(sendAccountReceive.created_at).toBeInstanceOf(Date)
    })

    it('should have correct user data', () => {
      expect(sendAccountReceive.from_user).not.toBeNull()
      expect(sendAccountReceive.to_user).not.toBeNull()
    })

    it('should have correct receive data', () => {
      expect(sendAccountReceive.data.sender).toBe('0xa0Ee7A142d267C1f36714E4a8F75612F20a79720')
      expect(sendAccountReceive.data.value).toBe(10000000000000000n)
    })
  })

  describe('SendTokenUpgradeEvent', () => {
    let upgrade: SendAccountTransfersEvent

    beforeAll(() => {
      upgrade = parsedData[9] as SendAccountTransfersEvent
      assert(!!upgrade)
    })

    it('should have correct event name and timestamps', () => {
      expect(upgrade.event_name).toBe('send_account_transfers')
      expect(upgrade.created_at).toBeInstanceOf(Date)
    })

    it('should have correct user data', () => {
      expect(upgrade.from_user).toBeNull()
      expect(upgrade.to_user).not.toBeNull()
    })

    it('should have correct upgrade data', () => {
      expect(upgrade.data.f).toBe('0x0000000000000000000000000000000000000000')
      expect(upgrade.data.t).toBe('0x649667EfCcf6497290616e7A669024ffEAF75968')
      expect(upgrade.data.v).toBe(10000000000000000000000n)
      expect(upgrade.data.block_num).toBe(25254355n)
      expect(upgrade.data.tx_idx).toBe(0n)
      expect(upgrade.data.log_idx).toBe(5n)
      expect(upgrade.data.tx_hash).toBe(
        '0xc8e94001e225e3d4570c352a3811de04586c1cfabc8b7c9367d477fcf003424d'
      )
      expect(upgrade.data.coin).toEqual(sendCoin)
    })
  })
})
