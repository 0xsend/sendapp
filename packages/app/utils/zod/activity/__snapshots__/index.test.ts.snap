// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventArraySchema should parse a valid event array 1`] = `
[
  {
    "created_at": 2024-05-26T13:38:25.000Z,
    "data": {
      "block_num": 15164540n,
      "coin": {
        "coingeckoTokenId": "usd-coin",
        "decimals": 6,
        "formatDecimals": 2,
        "label": "USDC",
        "symbol": "USDC",
        "token": "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
      },
      "f": "0x760E2928C3aa3aF87897bE52eb4833d42bbB27cf",
      "log_addr": "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
      "log_idx": 2n,
      "t": "0xbf65EE06b43B9cA718216241f0b9F81b5ff30CC1",
      "tx_hash": "0xf1443b5abd14e6212dda2d9f5ff1d1d691599de3e8fa019ccc19b909d9bb46a4",
      "tx_idx": 0n,
      "v": 19032n,
    },
    "event_name": "send_account_transfers",
    "from_user": null,
    "to_user": {
      "avatar_url": null,
      "id": "97476407-bf7f-4ebe-86aa-c9d18a0b388a",
      "main_tag_id": 1,
      "main_tag_name": "asdf",
      "name": null,
      "send_id": 65244,
      "tags": [
        "asdf",
        "teq",
        "yuw",
      ],
    },
  },
  {
    "created_at": 2024-05-26T19:41:05.370Z,
    "data": {
      "block_num": 15164541n,
      "coin": {
        "coingeckoTokenId": "ethereum",
        "decimals": 18,
        "formatDecimals": 5,
        "label": "Ethereum",
        "symbol": "ETH",
        "token": "eth",
      },
      "log_addr": "0x71fa02bb11e4b119bEDbeeD2f119F62048245301",
      "log_idx": 2n,
      "tags": [
        "yuw",
      ],
      "tx_hash": "0x37c4281422413a3a78e765452c47abb5c3a95c102282bdd3632ced0b640d861c",
      "tx_idx": 0n,
      "value": 20000000000000000n,
    },
    "event_name": "tag_receipts",
    "from_user": {
      "avatar_url": null,
      "id": "97476407-bf7f-4ebe-86aa-c9d18a0b388a",
      "main_tag_id": 1,
      "main_tag_name": "asdf",
      "name": null,
      "send_id": 65244,
      "tags": [
        "asdf",
        "teq",
        "yuw",
      ],
    },
    "to_user": null,
  },
  {
    "created_at": 2024-05-26T19:41:05.370Z,
    "data": {
      "tags": [
        "disconnect_whorl7351",
      ],
    },
    "event_name": "referrals",
    "from_user": {
      "avatar_url": null,
      "id": "97476407-bf7f-4ebe-86aa-c9d18a0b388a",
      "main_tag_id": 1,
      "main_tag_name": "asdf",
      "name": null,
      "send_id": 65244,
      "tags": [
        "asdf",
        "teq",
        "yuw",
      ],
    },
    "to_user": {
      "avatar_url": "https://i.pravatar.cc/500?u=Hortense Ratke",
      "id": null,
      "main_tag_id": 2,
      "main_tag_name": "disconnect_whorl7351",
      "name": "Mr. Dan Conn",
      "send_id": 12530,
      "tags": [
        "disconnect_whorl7351",
      ],
    },
  },
  {
    "created_at": 2024-05-29T13:34:26.000Z,
    "data": {
      "account": "\\xa7ded3f6316c7d3b5ae2ed711cf535395db921b1",
      "block_num": "15164541",
      "key": [
        "\\x351631d94d8cfc12f6adfc2586111990681f216c7d6d8531e669471293f32f07",
        "\\x83577aa62079c3bb5b813017df43832562d133feb3a7447d28849dac74c8aa43",
      ],
      "key_slot": 0,
      "log_addr": "\\xa7ded3f6316c7d3b5ae2ed711cf535395db921b1",
      "log_idx": "2",
      "tx_hash": "\\x8c5eeeaf3c97bc2378854be36a99536dd181a1273b359ea714b9c1d4ed6c1e85",
      "tx_idx": "0",
    },
    "event_name": "send_account_signing_key_added",
    "from_user": {
      "avatar_url": null,
      "id": "97476407-bf7f-4ebe-86aa-c9d18a0b388a",
      "main_tag_id": null,
      "main_tag_name": null,
      "name": null,
      "send_id": 55617,
      "tags": null,
    },
    "to_user": null,
  },
  {
    "created_at": 2024-05-29T13:34:26.000Z,
    "data": {
      "account": "\\xa7ded3f6316c7d3b5ae2ed711cf535395db921b1",
      "block_num": "15164542",
      "key": [
        "\\x351631d94d8cfc12f6adfc2586111990681f216c7d6d8531e669471293f32f07",
        "\\x83577aa62079c3bb5b813017df43832562d133feb3a7447d28849dac74c8aa43",
      ],
      "key_slot": 0,
      "log_addr": "\\xa7ded3f6316c7d3b5ae2ed711cf535395db921b1",
      "log_idx": "2",
      "tx_hash": "\\x8c5eeeaf3c97bc2378854be36a99536dd181a1273b359ea714b9c1d4ed6c1e85",
      "tx_idx": "0",
    },
    "event_name": "send_account_signing_key_removed",
    "from_user": {
      "avatar_url": null,
      "id": "97476407-bf7f-4ebe-86aa-c9d18a0b388a",
      "main_tag_id": null,
      "main_tag_name": null,
      "name": null,
      "send_id": 55617,
      "tags": null,
    },
    "to_user": null,
  },
  {
    "created_at": 2024-05-27T13:36:07.156Z,
    "data": {
      "block_num": 15164541n,
      "coin": {
        "coingeckoTokenId": "usd-coin",
        "decimals": 6,
        "formatDecimals": 2,
        "label": "USDC",
        "symbol": "USDC",
        "token": "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
      },
      "f": "0x760E2928C3aa3aF87897bE52eb4833d42bbB27cf",
      "log_addr": "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
      "log_idx": 2n,
      "note": "Send gonna be $1 someday",
      "t": "0xbf65EE06b43B9cA718216241f0b9F81b5ff30CC1",
      "tx_hash": "0xf1443b5abd14e6212dda2d9f5ff1d1d691599de3e8fa019ccc19b909d9bb46a4",
      "tx_idx": 0n,
      "v": 77777n,
    },
    "event_name": "send_account_transfers",
    "from_user": {
      "avatar_url": null,
      "id": "97476407-bf7f-4ebe-86aa-c9d18a0b388a",
      "main_tag_id": 1,
      "main_tag_name": "asdf",
      "name": null,
      "send_id": 65244,
      "tags": [
        "asdf",
        "teq",
        "yuw",
      ],
    },
    "to_user": {
      "avatar_url": "https://i.pravatar.cc/500?u=Hortense Ratke",
      "id": null,
      "main_tag_id": 1,
      "main_tag_name": "dan",
      "name": "Mr. Dan Conn",
      "send_id": 12530,
      "tags": [
        "dan",
      ],
    },
  },
  {
    "created_at": 2024-06-12T00:31:48.000Z,
    "data": {
      "block_num": 15681483n,
      "coin": {
        "coingeckoTokenId": "ethereum",
        "decimals": 18,
        "formatDecimals": 5,
        "label": "Ethereum",
        "symbol": "ETH",
        "token": "eth",
      },
      "log_addr": "0xB2c21F54653531aa4AffA80F63593913f0C70628",
      "log_idx": 0n,
      "note": "Send gonna be $1 someday",
      "sender": "0xa0Ee7A142d267C1f36714E4a8F75612F20a79720",
      "tx_hash": "0xeec33cc50042cbba53fc1de714bd99b206635f890dbe29771c7986df6da0f6af",
      "tx_idx": 0n,
      "value": 10000000000000000n,
    },
    "event_name": "send_account_receives",
    "from_user": {
      "avatar_url": null,
      "id": null,
      "main_tag_id": null,
      "main_tag_name": null,
      "name": null,
      "send_id": 65244,
      "tags": null,
    },
    "to_user": {
      "avatar_url": null,
      "id": "97476407-bf7f-4ebe-86aa-c9d18a0b388a",
      "main_tag_id": 1,
      "main_tag_name": "asdf",
      "name": null,
      "send_id": 65244,
      "tags": [
        "asdf",
        "teq",
        "yuw",
      ],
    },
  },
  {
    "created_at": 2024-05-26T19:41:05.370Z,
    "data": {
      "block_num": 15164541n,
      "coin": {
        "coingeckoTokenId": "usd-coin",
        "decimals": 6,
        "formatDecimals": 2,
        "label": "USDC",
        "symbol": "USDC",
        "token": "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
      },
      "log_addr": "0x71fa02bb11e4b119bEDbeeD2f119F62048245301",
      "log_idx": 2n,
      "tags": [
        "tag_receipt_usdc",
      ],
      "tx_hash": "0x37c4281422413a3a78e765452c47abb5c3a95c102282bdd3632ced0b640d861c",
      "tx_idx": 0n,
      "value": 2000000n,
    },
    "event_name": "tag_receipt_usdc",
    "from_user": {
      "avatar_url": null,
      "id": "97476407-bf7f-4ebe-86aa-c9d18a0b388a",
      "main_tag_id": 1,
      "main_tag_name": "asdf",
      "name": null,
      "send_id": 65244,
      "tags": [
        "asdf",
        "teq",
      ],
    },
    "to_user": null,
  },
  {
    "created_at": 2024-07-18T19:32:08.000Z,
    "data": {
      "block_num": 17270890n,
      "coin": {
        "coingeckoTokenId": "usd-coin",
        "decimals": 6,
        "formatDecimals": 2,
        "label": "USDC",
        "symbol": "USDC",
        "token": "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
      },
      "f": "0xfC1e51BBae1C1Ee9e6Cc629ea0023329EA5023a6",
      "log_addr": "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
      "log_idx": 4n,
      "t": "0x29369A06d3c17897E819C43202A7f4A1004B20cF",
      "tx_hash": "0x797a6a0fdd1c849e4e6e58bdc02966d19cff1eb8ef257216914288e715e460b8",
      "tx_idx": 0n,
      "v": 8000000n,
    },
    "event_name": "send_account_transfers",
    "from_user": null,
    "to_user": {
      "avatar_url": null,
      "id": "5def5ec4-7f0e-4509-990f-d52e85e95187",
      "main_tag_id": null,
      "main_tag_name": null,
      "name": null,
      "send_id": 64705,
      "tags": null,
    },
  },
  {
    "created_at": 2024-05-29T14:47:36.000Z,
    "data": {
      "block_num": 25254355n,
      "coin": {
        "coingeckoTokenId": "send-token-2",
        "decimals": 18,
        "formatDecimals": 0,
        "label": "Send",
        "symbol": "SEND",
        "token": "0xEab49138BA2Ea6dd776220fE26b7b8E446638956",
      },
      "f": "0x0000000000000000000000000000000000000000",
      "log_addr": "0xEab49138BA2Ea6dd776220fE26b7b8E446638956",
      "log_idx": 5n,
      "t": "0x649667EfCcf6497290616e7A669024ffEAF75968",
      "tx_hash": "0xc8e94001e225e3d4570c352a3811de04586c1cfabc8b7c9367d477fcf003424d",
      "tx_idx": 0n,
      "v": 10000000000000000000000n,
    },
    "event_name": "send_account_transfers",
    "from_user": null,
    "to_user": {
      "avatar_url": null,
      "id": "5c833c74-5a6c-461d-8612-b4982d7a7b51",
      "main_tag_id": 1,
      "main_tag_name": "asfd",
      "name": null,
      "send_id": 48,
      "tags": [
        "asfd",
      ],
    },
  },
  {
    "created_at": 2025-01-29T14:47:36.000Z,
    "data": {
      "assets": 1000000000000000000n,
      "block_num": 12345n,
      "log_addr": "0x5AFE000000000000000000000000000000000000",
      "log_idx": 0n,
      "owner": "0xB0b0000000000000000000000000000000000000",
      "sender": "0xB0b0000000000000000000000000000000000000",
      "shares": 950000000000000000n,
      "tx_hash": "0x1234567890123456789012345678901234567890123456789012345678901234",
      "tx_idx": 0n,
    },
    "event_name": "send_earn_deposit",
    "from_user": {
      "avatar_url": null,
      "id": "5c833c74-5a6c-461d-8612-b4982d7a7b51",
      "main_tag_id": 1,
      "main_tag_name": "asfd",
      "name": null,
      "send_id": 48,
      "tags": [
        "asfd",
      ],
    },
    "to_user": null,
  },
  {
    "created_at": 2025-02-15T14:47:36.000Z,
    "data": {
      "assets": 5000000000000000000n,
      "block_num": 12345n,
      "log_addr": "0x5AFE000000000000000000000000000000000000",
      "log_idx": 0n,
      "owner": "0xB0b0000000000000000000000000000000000000",
      "receiver": "0xB0b0000000000000000000000000000000000000",
      "sender": "0xB0b0000000000000000000000000000000000000",
      "shares": 475000000000000000n,
      "tx_hash": "0x1234567890123456789012345678901234567890123456789012345678901234",
      "tx_idx": 0n,
    },
    "event_name": "send_earn_withdraw",
    "from_user": null,
    "to_user": {
      "avatar_url": null,
      "id": "5c833c74-5a6c-461d-8612-b4982d7a7b51",
      "main_tag_id": 1,
      "main_tag_name": "asfd",
      "name": null,
      "send_id": 48,
      "tags": [
        "asfd",
      ],
    },
  },
]
`;
